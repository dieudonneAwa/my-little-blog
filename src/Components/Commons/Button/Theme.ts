import { css } from "styled-components";
import { darken, lighten, transparentize } from "polished";
import { font, space } from "../../../Theme/typography";

// BASE
const buttonBase = css`
  font-weight: 500;
  transition: all 0.2s linear;
  min-width: fit-content;
  font-size: ${font.size.medium};

  &:hover {
    cursor: pointer;
    transform: scale(1);
    transition: all 0.5s cubic-bezier(0.075, 0.82, 0.165, 1);
  }
  &:not(:hover) {
    transform: scale(1);
    transition: all 0.6s cubic-bezier(0.075, 0.82, 0.165, 1);
  }
  @media (max-width: 767px) {
    padding: ${space.small};
  }
`;

// COLOR

const emerald = css`
  background: ${(p) => p.theme.colors.emerald};
  color: ${(p) => p.theme.colors.white};
  box-shadow: -1px -1px 8px ${(p) => transparentize(0.7, darken(0.2, p.theme.colors.emerald))};
  border: 1px solid ${(p) => p.theme.colors.emerald};

  &:hover {
    background: ${(p) => lighten(0.05, p.theme.colors.emerald)};
    box-shadow: 0px 0px 12px
      ${(p) => transparentize(0.7, darken(0.2, p.theme.colors.emerald))};
  }
`;

const spacecadet = css`
  border: 1px solid ${(p) => p.theme.colors.spacecadet};
  background: ${(p) => p.theme.colors.spacecadet};
  color: ${(p) => p.theme.colors.white};

  &:hover {
    background: ${(p) => darken(0.05, p.theme.colors.spacecadet)};
    box-shadow: 0px 0px 12px
      ${(p) => transparentize(0.7, darken(0.2, p.theme.colors.spacecadet))};
  }
`;

const blueviolet = css`
  border: 1px solid ${(p) => p.theme.colors.blueviolet};
  background: ${(p) => p.theme.colors.blueviolet};
  color: ${(p) => p.theme.colors.white};

  &:hover {
    background: ${(p) => darken(0.05, p.theme.colors.blueviolet)};
    box-shadow: 0px 0px 12px
      ${(p) => transparentize(0.7, darken(0.2, p.theme.colors.blueviolet))};
  }
`;

const defaultcolor = css``;

const buttonColor: any = {
  emerald,
  spacecadet,
  default: defaultcolor,
  blueviolet,
};

// SIZE

const small = css`
  font-size: ${font.size.medium};
  padding: ${space.tiny} ${space.medium};
  border-radius: ${(p) => p.theme.radii.small};
`;

const medium = css`
  font-size: ${font.size.medium};
  padding: ${space.small} ${space.large};
  border-radius: ${(p) => p.theme.radii.small};
`;

const large = css`
  font-size: ${font.size.large};
  padding: ${space.medium} ${space.large};
  border-radius: ${(p) => p.theme.radii.medium};
`;

const buttonSize: any = {
  small,
  medium,
  large,
};

// SHAPE
const rounded = css``;

const square = css`
  border-radius: 0;
`;
const pill = css`
  border-radius: 50px;
`;

const buttonShape: any = {
  rounded,
  square,
  pill,
};

export { buttonBase, buttonColor, buttonSize, buttonShape };
