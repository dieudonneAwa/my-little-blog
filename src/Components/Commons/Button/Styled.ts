import styled from 'styled-components'
import { space } from 'styled-system'
import { buttonBase, buttonColor, buttonSize, buttonShape } from './Theme'

export const StyledButton = styled.button<{
  color: string
  size: string
  shape: string
}>`
  ${buttonBase}
  ${(p) => buttonColor[p.color]};
  ${(p) => buttonSize[p.size]};
  ${(p) => buttonShape[p.shape]};

  ${space}
`
