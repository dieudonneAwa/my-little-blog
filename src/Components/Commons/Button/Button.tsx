/* eslint-disable react/forbid-prop-types */
import React from "react";
import { StyledButton } from "./Styled";
import { Spinner } from "../Spinner";
import { SpaceProps } from "styled-system";

export type Props = SpaceProps & {
  /** Defines color of the button based on theme */
  color?: "emerald" | "spacecadet" | "blueviolet" | "default";

  /** This is the content of the button. Detault: 'Button' */
  children?: string | Node;

  /** The size of the button. Default: 'small' */
  size?: "small" | "medium" | "large";

  /** Shows a loading button. Default: false */
  loading?: boolean;

  /** Shape of the button through absolute radii */
  shape?: "rounded" | "pill" | "square";

  /** onClickHandler */
  onClick?: (e?: any) => void;

  m?: string;
};

const Button = ({
  color = "emerald",
  children = "button",
  onClick,
  size = "small",
  shape = "rounded",
  loading = false,
  ...rest
}: Props) => {
  const content = loading ? (
    <>
      {children}
      <Spinner ml="small" color="l001" size="0.6rem" />
    </>
  ) : (
    children
  );
  return (
    <StyledButton
      type="button"
      color={color}
      size={size}
      shape={shape}
      onClick={onClick}
      m="0.2rem"
      {...rest}
    >
      {content}
    </StyledButton>
  );
};

export default Button;
