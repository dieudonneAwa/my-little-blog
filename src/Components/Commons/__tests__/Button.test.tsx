/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Button from '../Button/Button';

describe('Button', (): void => {
  let wrapper: ShallowWrapper<void>;
  beforeEach((): void => {
    wrapper = shallow(<Button />);
  });

  it('should render Button component', (): void => {
    wrapper.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
});
