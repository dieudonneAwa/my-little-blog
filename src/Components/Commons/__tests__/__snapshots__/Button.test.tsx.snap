// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render Button component 1`] = `
<styled.button
  color="emerald"
  m="0.2rem"
  shape="rounded"
  size="small"
  type="button"
>
  button
</styled.button>
`;
